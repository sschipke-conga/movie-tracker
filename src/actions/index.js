export const setMovies = (movies) => ({
  type: 'SET_MOVIES',
  movies,
});

export const setUpcomingMovies = (upcomingMovies) => ({
  type: 'SET_UPCOMING_MOVIES',
  upcomingMovies,
});

export const setUser = (user) => ({
  type: 'SET_USER',
  user,
});

export const setFavorites = (favorites) => ({
  type: 'SET_FAVORITES',
  favorites,
});
