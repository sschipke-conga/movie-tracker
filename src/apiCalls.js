export const getMovies = async () => {
  let res = await fetch('https://api.themoviedb.org/3/discover/movie?sort_by=popularity.desc&api_key=9e1950f701d0af712babdede31811e9e')
  let data = await res.json()
  return data.results
}