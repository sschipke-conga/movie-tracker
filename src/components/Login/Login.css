.screen-cover {
  position:absolute;
  background:rgba(0, 0, 0, 0.8);
  width: 100%;
  height:100vh;

}

.login {
  display: flex;
  justify-content: center;
}


.login__form {
  background: url(../../images/form-background.jpg);
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 100%;
}

input {
  font-size: 20px;
  height: 35px;
  line-height: 30px;
  padding:0 8px;
  width:250px;
  border-radius:5px;
  margin:20px 0;
}

input::placeholder {
  color:white;
  font-size: 15px;
}

.login__btn {
  font-size: 1em;
  height:30px;
  margin-bottom: 20px;
  border-radius: 5px;
  width: 250px;
}

.login__btn:hover {
  cursor: pointer;
  background: red;
  color: white;
}

.error_login {
  color: red;
}

.error {
  border: 2px solid red;
}