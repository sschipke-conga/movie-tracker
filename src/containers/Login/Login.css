.screen-cover {
  position:absolute;
  background:rgba(0, 0, 0, 0.8);
  width: 100%;
  height:100vh;
}

.login {
  display: flex;
  justify-content: center;
  align-items: center;
  background: url(../../images/form-background.jpg);
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
}

.login__form {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 800px;
  height:450px;

  border:1px solid rgba(255, 255, 255, 0.5);
  background-color:rgba(0,0,0,0.8)
}

input {
  font-size: 20px;
  height: 35px;
  line-height: 30px;
  padding:0 8px;
  width:250px;
  border-radius:5px;
  margin:20px 0;
}

input::placeholder {
  color:white;
  font-size: 15px;
}

.login__btn {
  font-size: 1em;
  height:30px;
  margin-bottom: 20px;
  border-radius: 5px;
  width: 250px;
}

.login__btn:hover {
  cursor: pointer;
  background: red;
  color: white;
}

.error_login {
  color: red;
}

.error {
  border: 2px solid red;
}