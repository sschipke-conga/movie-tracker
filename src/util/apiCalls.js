export const getMovies = async () => {
  let res = await fetch('https://api.themoviedb.org/3/movie/now_playing?api_key=9e1950f701d0af712babdede31811e9e&language=en-US')
  let data = await res.json()
  return cleanMovieData(data.results)
}

const cleanMovieData = data => data.map(movie => ({
    poster_path: movie.poster_path,
    title: movie.title,
    release_date: movie.release_date,
    vote_average: movie.vote_average,
    overview: movie.overview,
    movie_id: movie.id
  }))

export const getUpcomingMovies = async () => {
  let res = await fetch('https://api.themoviedb.org/3/movie/upcoming?api_key=9e1950f701d0af712babdede31811e9e&language=en-US')
  let data = await res.json()
  return cleanMovieData(data.results)
}

export const createNewUser = async user => {
  let url = 'http://localhost:3001/api/v1/users'
  let options = {
      method: 'POST',
      body: JSON.stringify(user),
        headers: {
      'Content-Type': 'application/json'
    }
  }
  let res = await fetch(url, options);

  if (!res.ok) {
    if(res.status===500) {
      throw Error('This email has already been used')
    }
  }
  return res.json();
}

export const logInUser = async user => {
  let url = 'http://localhost:3001/api/v1/login';
  let options = {
    method: 'POST',
    body: JSON.stringify(user),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  let res = await fetch(url, options)

  if (!res.ok) {
    if (res.status === 401) {
      throw Error('email or password is incorrect')
    }
  }
  return res.json();
}

export const getUserFavorites = async userID => {
  let url = `http://localhost:3001/api/v1/users/${userID}/moviefavorites`;
  let res = await fetch(url);
  let parsedRes = await res.json();
  return parsedRes.favorites;
}

export const postFavorite = async (userId, movie) => {
  console.log('post', userId, movie)
  let url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites`;
  let options = {
    method: 'POST',
    body: JSON.stringify(movie),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  let res = await fetch(url, options);
  if (!res.ok) {
    throw Error('Could not add favorite movie.')
  }
}


export const deleteFavorite = async (userId, movie_id) => {
  console.log('delete', userId, movie_id)
  let url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites/${movie_id}`
  let options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }
  let res = await fetch(url, options);
  if (!res.ok) {
      throw Error('Could not delete favorite.')
  }

}
