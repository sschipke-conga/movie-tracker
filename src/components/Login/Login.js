import React,{ Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { Redirect } from 'react-router-dom'
import { createNewUser, logInUser , getUserFavorites} from '../../util/apiCalls';
import {setUser, setFavorites} from '../../actions';
import './Login.css';

export class Login extends Component {
  constructor(){
    super()
    this.state={
      name:'',
      email:'',
      password:'',
      isLoggedIn:false,
      error: '',
      logInError: false
    }
  }

  handleChange = (e) => {
    this.setState( { [e.target.name]:e.target.value } )
  }

  handleSubmit = async (e) => {
    e.preventDefault()
    this.createUser()
    }
    


  createUser = async () => {
    const { name, email, password } = this.state;
    const user = {
      name,
      email,
      password
    }
    try {
      let newUser = await createNewUser(user)
      await this.props.setUser(newUser)
      this.setState({ isLoggedIn: true })
    } catch ({ message }) {
      this.setState({ error: message })
    }
  }

  logIn = async () => {
    const { email, password } = this.state;
    let user = { email, password }
    try {
      let currentUser = await logInUser(user);
      let userFavorites = await getUserFavorites(currentUser.id);
      await this.props.setFavorites(userFavorites)
      await this.props.setUser(currentUser)
      this.setState({isLoggedIn: true})
    } catch ({ message }) {
      this.setState({ error: message, logInError:true })
    }
  }

  render = () => {
    const {isLoggedIn, error, logInError} = this.state;
    let errClass = logInError ? 'error' : ''
    if(isLoggedIn) {
      return <Redirect to='/' />
    }
    return(
      <div className='login screen-cover'>
        <form className='login__form' onSubmit={this.handleSubmit}>

          <input 
            placeholder='Name' 
            name='name'
            maxLength='15'
            required 
            value={this.state.name}
            onChange={this.handleChange}/>

          <input 
          type='email'
            placeholder='Email' 
            className={errClass}
            name='email' 
            required
            value={this.state.email}
            onChange={this.handleChange}/>

          <input 
            type='password' 
            placeholder='Password' 
            className={errClass}
            name='password'
            maxLength='20'
            required 
            value={this.state.password}
            onChange={this.handleChange}/>
          <button type='submit'
          id='create-new'
          className='login__btn'
          >Create an account</button>
          <button type='button' className='login__btn'
            id='login'
            onClick={this.logIn}
            >Login</button>
            {error && <h3 className='error__login'>{error}</h3>}
        </form>
      </div>
    )
  }
}

export const mapDispatchToProps = dispatch => bindActionCreators({setUser, setFavorites}, dispatch)


export default connect(null, mapDispatchToProps)(Login);