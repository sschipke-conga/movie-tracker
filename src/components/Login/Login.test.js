import React from 'react';
import { Login, mapStateToProps } from './Login';
import { shallow } from 'enzyme';
import { getUserFavorites } from '../../util/apiCalls';
import { setFavorites } from '../../actions';

jest.mock('../../util/apiCalls');

describe('Login', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Login />)
  })
  it('should match the initial snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  });
  describe('handleChange', () => {
    it('should update state if a user types in their name', () => {
      const mockEvent = {target: {name: 'name', value: 'Steve' }};
      expect(wrapper.state('name')).toEqual('');
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('name')).toEqual(mockEvent.target.value)
    })
    it('should update state if a user types in their email', () => {
      const mockEvent = { target: { name: 'email', value: 'steve@g.com' } };
      wrapper.instance().setState({name: 'Steve'});
      expect(wrapper.state('email')).toEqual('');
      expect(wrapper.state('name')).toEqual('Steve')

      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('email')).toEqual(mockEvent.target.value)
    expect(wrapper.state('name')).toEqual('Steve')
    })
    it('should update state if a user types in their password', () => {
      const mockEvent = { target: { name: 'password', value: 'password' } };
      wrapper.instance().setState({ name: 'Steve' });
      wrapper.instance().setState({ email: 'steve@g.com' });
      expect(wrapper.state('email')).toEqual('steve@g.com');
      expect(wrapper.state('name')).toEqual('Steve')

      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('password')).toEqual(mockEvent.target.value);
      expect(wrapper.state('name')).toEqual('Steve');
      expect(wrapper.state('email')).toEqual('steve@g.com')
    })
  })
  describe('handleSubmit', () => {
    
  })
})
