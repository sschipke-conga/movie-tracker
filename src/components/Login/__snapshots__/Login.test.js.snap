// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match the initial snapshot 1`] = `
<div
  className="login screen-cover"
>
  <form
    className="login__form"
    onSubmit={[Function]}
  >
    <input
      maxLength="15"
      name="name"
      onChange={[Function]}
      placeholder="Name"
      required={true}
      value=""
    />
    <input
      className=""
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required={true}
      type="email"
      value=""
    />
    <input
      className=""
      maxLength="20"
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
      value=""
    />
    <button
      className="login__btn"
      id="create-new"
      type="submit"
    >
      Create an account
    </button>
    <button
      className="login__btn"
      id="login"
      onClick={[Function]}
      type="button"
    >
      Login
    </button>
  </form>
</div>
`;
